rfcs/20190502-preprocessing-layers.md

```diff
@@ -0,0 +1,578 @@
+
+# Map the preprocessing stage on the dataset.
+dataset = Dataset(...)
+dataset = dataset.map(preproc_stage)
+
+# Regular training loop (using `main_model`).
+for x, y in dataset:
+    with GradientTape() as tape:
+        y_pred = main_model(x)
+        loss = loss_fn(y, y_pred)
+        ...
+```
+
+In general, you won't have to refer to `get_preprocessing_stage` and `get_main_stage` directly, because you will
+probably already have direct handles on your preprocessing layer and the rest of the model:
+
+```python
+normalization = layers.Normalization()
+normalization.adapt(preproc_dataset)
+dataset = dataset.map(normalization)
+
+for x, y in dataset:
+    with GradientTape() as tape:
+        y_pred = model(x)
+        loss = loss_fn(y, y_pred)
+        ...
+```
+
+
+## Questions and Discussion Topics
+
+### Naming Discussion
+
+#### Naming conventions to follow for preprocessing layers
+
+We have two possible sets of names for the layers:
+
+##### Option A: Normalization, Discretization, TextVectorization
+
+Pros: consistent with most existing layers, in particular BatchNormalization.
+Cons: It's longer.
+
+##### Option B: Normalize, Discretize, VectorizeText
+
+Pros: It's shorter.
+Cons: Normalize vs BatchNormalization is jarring.
+
+
+#### Using the name "preprocessing" or "processing"
+
+It has been proposed that we use the name "processing" throughout the API instead of "preprocessing".
+
+##### Option A: "preprocessing".
+
+Pros:
+1) The meaning of "preprocessing" is clear to all users ("data normalization and stuff").
+2) We need a clear semantic boundary between the main data processing flow of a model and what goes before it (the preprocessing stage).
+3) It replaces the functionality of the `keras.preprocessing` module, and should be consistent with this naming convention.
+
+Cons:
+The `Normalization` layer, being differentiable, can be used in the middle of a model, rather than at the start.
+However, there's nothing weird about keeping the name "preprocessing" in this specific case: it is widely understood that a `Normalization` layer is doing "data preprocessing", independently of where you use it -- in fact, normalization is the first example that shows up in most definitions of "data preprocessing". 
+
+
+##### Option B: "processing".
+
+Pros: The Normalization layer can be used elsewhere in a model than at the start (although it would have to be trained separately).
+Cons: It's very generic, and does not clearly convey the difference between "preprocessing stage" and "main processing stage" required by the async prefetching API.
+
+
+#### Name to use for `adapt` method
+
+We may want to use the name `fit` instead (other suggestions welcome).
+
+Pros of using `fit`: consistency with `model.fit()`, and the `fit` method on `ImageDataGenerator` and `Tokenizer` from the `keras.preprocessing` module.
+Cons of using `fit`: It may confuse users, since `preprocessing_layer.fit()` would have a different signature.
+
```

Comment:
Can we add a point here about what the Scikit-learn API does? I believe consistency with existing libraries APIs is important and avoids documentation lookups. That would be in the pros.

Reply:
The equivalent Scikit-Learn classes (called `*Vectorizer`) use `fit`, e.g.  [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html) -- despite having a different signature compared to `model.fit()`.

This would be an argument in favor of using `fit` here as well.

Reply:
The word `fit` is more appropriate IMO as it is consistent with other libraries as well. Though the `argument` that it can confuse some users is fair but I believe most of the users(not including beginners here) would be able to well differentiate between the two. For example, If `fit` is used on `preprocessing`, then you know it is data preprocessing and have extra things which `model.fit()` may or may not have in common